client/fileutils.c

```diff
@@ -649,6 +649,84 @@ int loadFileDICTIONARY(const char *preferredName, void *data, size_t *datalen, u
     return retval;
 }
 
+int loadFileDICTIONARY_safe(const char *preferredName, void **pdata, uint8_t keylen, uint16_t *keycnt) {
+
+    int block_size = 512;
+    int allocation_size = block_size;
+    size_t counter = 0;
+    int retval = PM3_SUCCESS;
+    char *path;
+    if (searchFile(&path, DICTIONARIES_SUBDIR, preferredName, ".dic") != PM3_SUCCESS)
+        return PM3_EFILE;
+
+    // t5577 == 4bytes
+    // mifare == 6 bytes
+    // iclass == 8 bytes
+    // default to 6 bytes.
+    if (keylen != 4 && keylen != 6 && keylen != 8) {
+        keylen = 6;
+    }
+
+    // double up since its chars
+    keylen <<= 1;
+
+    char line[255];
+
+    // allocate some space for the dictionary
+    *pdata = calloc(keylen * allocation_size, sizeof(uint8_t));
+    if (*pdata == NULL) return PM3_EFILE;
+
+    FILE *f = fopen(path, "r");
+    if (!f) {
+        PrintAndLogEx(WARNING, "file not found or locked. '" _YELLOW_("%s")"'", path);
+        retval = PM3_EFILE;
+        goto out;    }
+
+    // read file
+    while (fgets(line, sizeof(line), f)) {
+        // check if we have enough space (if not allocate more)
+        if ((*keycnt) >= allocation_size) {
+            allocation_size += block_size;
+            *pdata = realloc(*pdata, keylen * allocation_size * sizeof(uint8_t));
+            if (*pdata == NULL) {
+                return PM3_EFILE;
+            } else {
+                // zero the new memeory (safety first)
+                memset(*pdata + counter, 0, block_size);
```

Comment:
is *pdata+counter really aligned on new mem?
I'd rather see *pdata+allocation_size-block_size

Reply:
If this isn't aligned, then this **num_to_bytes(key, keylen >> 1, *pdata + counter);**, would fail too, or wouldn't it?

Reply:
see when keylen=6, counter progress by +3 so it can't be aligned on blocksize (512).
when zeroizing mem, you need to zeroize the new 512b block.
In the code you show, it's just the regular injection of keys that doesn't care about blocksize alignment, no problem there

Reply:
Ahhh ok, now I see...